Src/PerformanceCollector/Filtering/CollectionConfiguration.cs

```diff
@@ -0,0 +1,331 @@
+
+        /// <summary>
+        /// Gets a list of performance counters.
+        /// </summary>
+        /// <remarks>
+        /// Performance counter name is stored in OperationalizedMetricInfo.Projection.
+        /// </remarks>
+        public IEnumerable<Tuple<string, string>> PerformanceCounters => this.performanceCounters;
+
+        private static void AddMetric<TTelemetry>(
+          OperationalizedMetricInfo metricInfo,
+          List<OperationalizedMetric<TTelemetry>> metrics,
+          out CollectionConfigurationError[] errors)
+        {
+            errors = new CollectionConfigurationError[] { };
+
+            try
+            {
+                metrics.Add(new OperationalizedMetric<TTelemetry>(metricInfo, out errors));
+            }
+            catch (Exception e)
+            {
+                // error creating the metric
+                errors =
+                    errors.Concat(
+                        new[]
+                        {
+                            CollectionConfigurationError.CreateError(
+                                CollectionConfigurationErrorType.MetricFailureToCreate,
+                                string.Format(CultureInfo.InvariantCulture, "Failed to create metric {0}.", metricInfo),
+                                e,
+                                Tuple.Create("MetricId", metricInfo.Id))
+                        }).ToArray();
+            }
+        }
+
+        private void CreatePerformanceCounters(out CollectionConfigurationError[] errors)
+        {
+            var errorList = new List<CollectionConfigurationError>();
+
+            OperationalizedMetricInfo[] performanceCounterMetrics =
+                (this.info.Metrics ?? new OperationalizedMetricInfo[0]).Where(metric => metric.TelemetryType == TelemetryType.PerformanceCounter)
+                    .ToArray();
+
+            this.performanceCounters.AddRange(
+                performanceCounterMetrics.GroupBy(metric => metric.Id, StringComparer.Ordinal)
+                    .Select(group => group.First())
+                    .Select(pc => Tuple.Create(pc.Id, pc.Projection)));
+
+            IEnumerable<string> duplicateMetricIds =
+                performanceCounterMetrics.GroupBy(pc => pc.Id, StringComparer.Ordinal).Where(group => group.Count() > 1).Select(group => group.Key);
+
+            foreach (var duplicateMetricId in duplicateMetricIds)
+            {
+                errorList.Add(
+                    CollectionConfigurationError.CreateError(
+                        CollectionConfigurationErrorType.PerformanceCounterDuplicateIds,
+                        string.Format(CultureInfo.InvariantCulture, "Duplicate performance counter id '{0}'", duplicateMetricId),
+                        null,
+                        Tuple.Create("MetricId", duplicateMetricId)));
+            }
+
+            errors = errorList.ToArray();
+        }
+
+        private void CreateDocumentStreams(
```

Comment:
See description for the DocumentStream class. I think whoever is looking at this private method implementation will get to that point, no need to copy that comment here.